src/uct/ib/rc/verbs/rc_verbs_iface.c

```diff
@@ -142,10 +143,22 @@ uct_rc_verbs_iface_poll_rx(uct_rc_verbs_iface_t *iface)
                                   wc[i].wr_id, wc[i].status);
             uct_ib_log_recv_completion(&iface->super.super, IBV_QPT_RC, &wc[i],
                                        hdr, uct_rc_ep_am_packet_dump);
-            uct_ib_iface_invoke_am(&iface->super.super, hdr->am_id, hdr + 1,
-                                   wc[i].byte_len - sizeof(*hdr), desc);
-        }
 
+            fc_hdr = uct_rc_ep_get_fc_hdr(hdr->am_id);
+
+            if (fc_hdr) {
+                hdr->am_id &= ~UCT_RC_EP_FC_MASK; /* Clear FC bits */
+                uct_rc_ep_handle_fc(&iface->super, wc[i].qp_num, fc_hdr);
+            }
+
+            /* we need to call AM handler only if it is NOT just pure FC grant */
+            if (ucs_likely(fc_hdr ^ UCT_RC_EP_FC_PURE_GRANT)) {
```

Comment:
can use `!(fc_hdr & UCT_RC_EP_FC_PURE_GRANT)` here, it's not better for performance to use xor, since at least on x86 we would like the compiler to emit "test" instruction anyway. Also, IMHO it's less readable this way.